protostar/starknet_gateway/gateway_facade.py

```diff
@@ -110,20 +115,27 @@ def _load_compiled_contract(self, compiled_contract_path: Path):
         except FileNotFoundError as err:
             raise CompilationOutputNotFoundException(compiled_contract_path) from err
 
-    def _transform_constructor_inputs_from_python(
+    def _prepare_constructor_inputs(
         self, inputs: Optional[CairoOrPythonData], compiled_contract_path: Path
-    ) -> List[int]:
-        if not inputs:
-            return []
-        if not isinstance(inputs, collections.Mapping):
-            return inputs
+    ):
+
         abi = self._compiled_contract_reader.load_abi_from_contract_path(
             compiled_contract_path
         )
         assert abi is not None
-        return from_python_transformer(abi, fn_name="constructor", mode="inputs")(
-            inputs
-        )
+
+        if not inputs:
+            return []
+        if not has_abi_item(abi, "constructor"):
+            if inputs:
+                raise InputValidationException(
+                    "Inputs provided to a contract with no constructor."
+                )
+        else:
+            inputs = transform_and_validate_constructor_inputs_from_python(abi, inputs)
+            validate_cairo_inputs(abi, inputs)
```

Comment:
```suggestion
        if not inputs:
            return []
            
        if inputs and not has_abi_item(abi, "constructor"):
            raise InputValidationException(
                "Inputs provided to a contract with no constructor."
            )

        inputs = transform_and_validate_constructor_inputs_from_python(abi, inputs)
        validate_cairo_inputs(abi, inputs)
```

this one is much more understandable at quick glance to me

Reply:
We need to prevent checking inputs to a constructor if it doesn't exist to avoid a LookupError.

Reply:
The only thing I did in this diff is restructuring code, I haven't changed the logic at all :)

Reply:
Sorry, `return []` is a bug that appeared during the refactor, it should be `inputs = []` (to also validate constructor with no inputs), when that's the case the logic is different.

Reply:
updated the snippet